danny-avila/LibreChat
# 💫 feat: Config File & Custom Endpoints

## Summary

Closes #1458
Closes #1344

I have extended the LibreChat app's flexibility by implementing custom endpoints, and introducing the `librechat.yaml` custom configuration file. 

This update allows users to define their own chat endpoints and have more control over the chat experience (for OpenAI Reverse proxies, such as OpenRouter, and Mistral, to name a few).

My updates include extensive backend/frontend API work to allow this change, with comprehensive updates to the documentation, improvements to existing logic. 

The motivation for these changes is to offer a more personalized and adaptable chat service that can cater to various user needs.

In future updates, some of the configurations from the env file will migrate to the new config file.

Further customization of the current configurations are also planned. Custom endpoints configurable from the frontend is also planned.

See `librechat.example.yaml` and `docs/install/configuration/custom_config.md` for setup

### Notable Improvements
-  fix(useSSE): handle errorHandler edge cases where error response is and is not properly formatted from API, especially when a conversationId is not yet provided, which ensures stream is properly closed on error https://github.com/danny-avila/LibreChat/pull/1474/commits/204cc1ed8a4471241a78afd504dbaeb994a55040
-  fix(streamResponse): handle error edge case where either messages or convos query throws an error https://github.com/danny-avila/LibreChat/pull/1474/commits/e3b7b065cb5344b79312ecc8b7afc8722a03eb5b

## Change Type

## Testing
Testing was conducted to ensure the custom endpoints function as expected. This involved:
- Verifying the correct loading and parsing of YAML configuration files.
- Testing the API's ability to handle custom endpoint requests.
- Confirming the frontend client can correctly display and interact with custom endpoints.
- Ensuring existing functionality remains unaffected by the new changes.

To reproduce my tests:
1. Set up custom endpoints in the `librechat.yaml` configuration file.
2. Run the server and interact with the application to verify the custom endpoints appear and function as intended.
3. Ensure that API responses are correctly formatted and displayed in the client.
4. Create various scenarios where custom endpoints are used to validate the robustness of the implementation.

### **Test Configuration**:
- Custom endpoints defined in `librechat.yaml`
- LibreChat app running locally for frontend interaction
- Backend API tested with Postman

## Checklist
